core/trino-main/src/main/java/io/trino/execution/FutureStateChange.java

```diff
@@ -44,8 +45,11 @@ public ListenableFuture<T> createNewListener()
         // remove the listener when the future completes
         listener.addListener(
                 () -> {
-                    synchronized (listeners) {
-                        listeners.remove(listener);
+                    // Only listeners that are canceled before being notified need individual removal
```

Comment:
Not sure I understand this. If the future completes normally, it should still be removed from the list to avoid "leaking" completed listeners. What am I missing?

Reply:
If a listener completes normally, it means that it has already been removed from the current listeners set since the set is copied and cleared before sending the notification in the first place. Only listeners that are canceled (via timeout, generally) need to acquire the lock to remove themselves from the current listeners.

Reply:
Got it. Please add that to the explanation. Otherwise it's too cryptic and one has to go digging to understand why that's the case.

Reply:
Done